### ISSUE TITLE
[Q] Can you share Softmax K experiment post / log?

### ISSUE THREAD

**Issue by apivovarov** [facebookincubator/AITemplate#350]
Softmax codegen for CUDA [backend/cuda/softmax/softmax.py](https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/backend/cuda/softmax/softmax.py#L16) says:

```
# For each K, whether to use wrapReduce or blockReduce was done by experiment
# Please refer to this post: https://fb.quip.com/HCfIAbpWB0qi
# and this experiment log: https://docs.google.com/spreadsheets/d/1bl3GCLQ67p27kXOSVJikEob38fojqaZIS--mPdQxeo0/edit#gid=931264442
```

Would it be possible to publish this post / experiment log to public?

**Comment by yinghai** [comment]
We are working on it.

**Comment by tissue3** [comment]
Here you go: https://github.com/facebookincubator/AITemplate/wiki/How-to-write-a-fast-Softmax-CUDA-kernel%3F

- The technique we used is block reduce, which is widely use to various block reduction operations in CUDA, e.g layernorm.
- The study shows that the performance is mostly  affected by pack_size at reading.
-  Because we applied a larger pack_size when possible, we were able to achieve better performance (50%) over oneflow.
- The doc also shows most codes we use to run the experiment and one should be able to reproduce the result with them.
